In [114]:
# import holder
from sqlalchemy import create_engine, text
import pandas as pd
import datetime as dt
import seaborn as sns

In [6]:
# bring in postgres database

database_name = 'scooters'
connection_string = f"postgresql://postgres:postgres@localhost:5432/{database_name}"
engine = create_engine(connection_string)

In [7]:
# header queries

scooters_sample = '''SELECT * FROM scooters LIMIT 10'''
trips_sample = '''SELECT * FROM trips limit 10'''

with engine.connect() as connection:
    scooters = pd.read_sql(text(scooters_sample), con = connection)
    trips = pd.read_sql(text(trips_sample), con = connection)

In [8]:
scooters.head()

pubdatetime   latitude  longitude                sumdid  \
0 2019-05-13 14:31:43.830  36.166379 -86.779052  PoweredY73AMEO5ZNOEN   
1 2019-05-13 14:31:43.830  36.164358 -86.781444  PoweredBP45IVHPZYERL   
2 2019-05-13 14:31:43.830  36.162163 -86.790782  Powered576Z23XBXXADT   
3 2019-05-13 14:31:43.830  36.165466 -86.781387  PoweredBBKCSVDXGGR3K   
4 2019-05-13 14:31:43.830  36.170682 -86.783562  Powered7UWCIA4TJK5BX   

  sumdtype  chargelevel sumdgroup  costpermin companyname  
0  Powered         87.0   scooter         0.0        Lime  
1  Powered         98.0   scooter         0.0        Lime  
2  Powered         98.0   scooter         0.0        Lime  
3  Powered         97.0   scooter         0.0        Lime  
4  Powered         98.0   scooter         0.0        Lime

In [9]:
trips.head()

pubtimestamp companyname triprecordnum         sumdid  \
0 2019-05-01 00:00:55.423        Bird       BRD2134   Powered9EAJL   
1 2019-05-01 00:03:33.147        Lyft          LFT5  Powered296631   
2 2019-05-01 00:05:55.570        Bird       BRD2168   Powered7S2UU   
3 2019-05-01 00:05:55.570        Bird       BRD2166   PoweredZIIVX   
4 2019-05-01 00:05:55.570        Bird       BRD2165   PoweredJ7MB3   

   tripduration  tripdistance   startdate        starttime     enddate  \
0        3.0000     958.00528  2019-05-01  00:00:20.460000  2019-05-01   
1        1.7156    1371.39112  2019-05-01  00:01:50.090000  2019-05-01   
2        3.0000    2296.58800  2019-05-01  00:03:47.363333  2019-05-01   
3        3.0000    1200.78744  2019-05-01  00:04:21.386666  2019-05-01   
4        2.0000     351.04988  2019-05-01  00:04:27.796666  2019-05-01   

           endtime  startlatitude  startlongitude  endlatitude  endlongitude  \
0  00:02:52.346666       36.15710       -86.80360     36.15660     -86.80670   
1  00:03:33.026666       36.15797       -86.77896     36.16054     -86.77689   
2  00:07:13.596666       36.15470       -86.78180     36.15650     -86.78680   
3  00:06:59.176666       36.14940       -86.77950     36.15310     -86.77960   
4  00:06:23.150000       36.17780       -86.78660     36.17740     -86.78760   

                                           triproute               create_dt  
0  [(36.157235, -86.803612), (36.157235, -86.8036... 2019-05-02 05:30:23.780  
1  [(36.15797, -86.77896), (36.15795, -86.77873),... 2019-05-02 07:20:32.757  
2  [(36.155068, -86.782124), (36.156597, -86.78675)] 2019-05-02 05:30:24.530  
3  [(36.149741, -86.779344), (36.149741, -86.7793... 2019-05-02 05:30:24.237  
4  [(36.177699, -86.786477), (36.177711, -86.7864... 2019-05-02 05:30:24.207

In [12]:
# parse scooter data to csvs. Uncomment lines below to write file to project directory; recomment when complete to avoid running query multiple times.
# Save the dataframe to a CSV file (example: df.to_csv('Biodata.csv', index=False))

#company queries
#bird_query = '''SELECT * FROM scooters WHERE companyname ILIKE 'bird';'''
#bolt_query = '''SELECT * FROM scooters WHERE companyname ilike 'bolt';'''
#gotcha_query = '''SELECT * FROM scooters WHERE companyname ilike 'gotcha';'''
#jump_query = '''SELECT * FROM scooters WHERE companyname ilike 'jump';'''
#lime_query = '''SELECT * FROM scooters WHERE companyname ilike 'lime' AND sumdgroup != 'bicycle';'''
#lyft_query = '''SELECT * FROM scooters WHERE companyname ilike 'lyft';'''
#spin_query = '''SELECT * FROM scooters WHERE companyname ilike 'spin';'''

# with engine.connect() as connection:
#    bird = pd.read_sql(text(bird_query), con = connection)
#     bolt = pd.read_sql(text(bolt_query), con = connection)
#     gotcha = pd.read_sql(text(gotcha_query), con = connection) 
#     jump = pd.read_sql(text(jump_query), con = connection)
#    lime = pd.read_sql(text(lime_query), con = connection)
#     lyft = pd.read_sql(text(lyft_query), con = connection)
#     spin = pd.read_sql(text(spin_query), con = connection)
    
#bird.to_csv('../data/scooters_bird.csv',index=False)
#bolt.to_csv('../data/scooters_bolt.csv',index=False)
#gotcha.to_csv('../data/scooters_gotcha.csv',index=False)
#jump.to_csv('../data/scooters_jump.csv',index=False)
#lime.to_csv('../data/scooters_lime.csv',index=False)
#lyft.to_csv('../data/scooters_lyft.csv',index=False)
#spin.to_csv('../data/scooters_spin.csv',index=False)


In [13]:
# parse trips data to csv.  Uncomment lines below to write file to project directory; recomment when complete to avoid running query multiple times.

# trips_query = '''SELECT * FROM trips WHERE sumdid NOT IN ('Standard5JXOV277MCWID', 'StandardNPOOZNUSGAXZN', 'StandardNUTLLXP4G37OI','StandardNW5HJFO4R32LY','StandardZPUQESHVPP74J');'''
# with engine.connect() as connection:
#      trips = pd.read_sql(text(trips_query), con = connection)
# trips.to_csv('../data/trips.csv',index=False)


In [14]:
#call in new csv dataframes

bird = pd.read_csv('../data/scooters_bird.csv')
bolt = pd.read_csv('../data/scooters_bolt.csv')
gotcha = pd.read_csv('../data/scooters_gotcha.csv')
jump = pd.read_csv('../data/scooters_jump.csv')
lyft = pd.read_csv('../data/scooters_lyft.csv')
spin = pd.read_csv('../data/scooters_spin.csv')

### EXPLORATION

#### EQ1: Are there any null values in any columns in either table?

In [15]:
# Use "company.info(verbose=True, show_counts=True)" to investicage shapes of parsed data

print("---BIRD INFO---")
print(bird.info(verbose=True, show_counts=True))
print("---BOLT INFO---")
print(bolt.info(verbbose=True, show_counts=True))
print("---GOTCHA INFO---")
print(gotcha.info(verbose=True, show_counts=True))
print("---JUMP INFO---")
print(jump.info(verbose=True, show_counts=True))
print("---LYFT INFO---")
print(lyft.info(verbose=True, show_counts=True))
print("---SPIN INFO---")
print(spin.info(verbose=True, show_counts=True))

## charge level: Bolt, Spin

---BIRD INFO---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12251590 entries, 0 to 12251589
Data columns (total 9 columns):
 #   Column       Non-Null Count     Dtype  
---  ------       --------------     -----  
 0   pubdatetime  12251590 non-null  object 
 1   latitude     12251590 non-null  float64
 2   longitude    12251590 non-null  float64
 3   sumdid       12251590 non-null  object 
 4   sumdtype     12251590 non-null  object 
 5   chargelevel  12251590 non-null  float64
 6   sumdgroup    12251590 non-null  object 
 7   costpermin   12251590 non-null  float64
 8   companyname  12251590 non-null  object 
dtypes: float64(4), object(5)
memory usage: 841.3+ MB
None
---BOLT INFO---


TypeError: DataFrame.info() got an unexpected keyword argument 'verbbose'

#### EA1. Bolt and Spin have missing some missing values in the chargelevel column

#### EQ2. What date range is represented in each of the date columns? Investigate any values that seem odd.

In [ ]:
# What date range is represented in each of the date columns? Investigate any values that seem odd.

#scooters - pubdatetime
scooters_date_query = '''SELECT pubdatetime FROM scooters'''
#trips - pubtimestamp, startdate, starttime, enddate, endtime, create_dt
trips_date_query = '''SELECT pubtimestamp, startdate, starttime, enddate, endtime, create_dt FROM trips'''

with engine.connect() as connection:
    scoot_dates = pd.read_sql(text(scooters_date_query), con = connection)
    trip_dates = pd.read_sql(text(trips_date_query), con = connection)
    
scoot_range_pubdate = scoot_dates.pubdatetime.max() - scoot_dates.pubdatetime.min()
trip_range_pubdate = trip_dates.pubtimestamp.max() - trip_dates.pubtimestamp.min()

In [16]:
# min_date
min_date = dates.dt.date.min()
print('The earliest date is: ' + str(min_date))
# max_date
max_date = dates.dt.date.max()
print('The latest date is: ' + str(max_date))
#range
print ('The date range is: ' + str(max_date - min_date))

NameError: name 'dates' is not defined

#### EA2.

#### EQ3. Is time represented with am/pm or using 24 hour values in each of the columns that include time?


In [17]:
#investigate max dates


#### EA3.

#### EQ4. What values are there in the sumdgroup column? Are there any that are not of interest for this project?

In [19]:
# What values are there in the sumdgroup column? Are there any that are not of interest for this project?
sumd_query = '''SELECT DISTINCT sumdgroup FROM scooters'''
with engine.connect() as connection:
    sumd_result = pd.read_sql(text(sumd_query), con = connection)
sumd_result

sumdgroup
0   bicycle
1   scooter
2   Scooter

In [20]:
# find bicycles 

bicycle_query = '''SELECT DISTINCT sumdid, companyname FROM scooters WHERE sumdgroup = 'bicycle';'''
with engine.connect() as connection:
    bikes = pd.read_sql(text(bicycle_query), con = connection)
bikes

sumdid companyname
0  Standard2UGJKREVB53HT        Lime
1  Standard5JXOV277MCWID        Lime
2  StandardNPOOZNUSGAXZN        Lime
3  StandardNUTLLXP4G37OI        Lime
4  StandardNW5HJFO4R32LY        Lime
5  StandardZPUQESHVPP74J        Lime

In [21]:
print(bikes.count())

sumdid         6
companyname    6
dtype: int64


#### EA4. The dataset contiains entires for six Lime bicycles; these entries should be omitted from scooter analysis

#### EQ5. What are the minimum and maximum values for all the latitude and longitude columns? Do these ranges make sense, or is there anything surprising?

In [ ]:
# What are the minimum and maximum values for all the latitude and longitude columns? Do these ranges make sense, or is there anything surprising? -What is the range of values for trip duration and trip distance? Do these values make sense? Explore values that might seem questionable.

sll_query = '''SELECT
                MIN(latitude) as min_latitude,
                MAX(latitude) as max_latitude,
                MIN(longitude) as min_longitude,
                MAX(latitude) as max_longitude
            FROM scooters'''
            
tll_query = '''SELECT
                MIN(startlatitude) as min_latitude_start,
                MAX(startlatitude) as max_latitude_start,
                MIN(startlongitude) as min_longitude_start,
                MAX(startlongitude) as max_longitude_start,
                MIN(endlatitude) as min_latitude_end,
                MAX(endlatitude) as max_latitude_end,
                MIN(endlongitude) as min_longitude_end,
                MAX(endlongitude) as max_longitude_end
            FROM trips'''

with engine.connect() as connection:
    scoot_result = pd.read_sql(text(sll_query), con = connection)
    trip_result = pd.read_sql(text(tll_query), con = connection)
    


In [ ]:
scoot_result

In [ ]:
trip_result

#### EA5.  The scooters table has a minimum latitude value of 0.0.  Latitude and longitude are expressed in defferent formats in the scooters and trips tables.

#### EQ6. Check out how the values for the company name column in the scooters table compare to those of the trips table. What do you notice?

In [ ]:
# Check out how the values for the company name column in the scooters table compare to those of the trips table. What do you notice?
scooter_companies = '''SELECT DISTINCT companyname FROM scooters'''
trips_companies = ''' SELECT DISTINCT companyname FROM trips'''

with engine.connect() as connection:
    scompanies = pd.read_sql(text(scooter_companies), con = connection)
    tcompanies = pd.read_sql(text(trips_companies), con = connection)

In [ ]:
scompanies.companyname.tolist()

In [ ]:
tcompanies.companyname.tolist()

#### EA6. The same companies are present in the scooters and trips tables, however there are some variations in capitalization.  Scooters contains company name "Bolt"; trips contains company name "Bolt Mobility"

Scooters: ['Bird', 'Bolt', 'Gotcha', 'Jump', 'Lime', 'Lyft', 'Spin']

Trips: ['Bird', 'Bolt Mobility', 'Gotcha', 'JUMP', 'Lime', 'Lyft', 'SPIN']

## ANALYSIS

#### Questions for consideration:
1. During this period, seven companies offered scooters. How many scooters did each company have in this time frame? Did the number for each company change over time? Did scooter usage vary by company?
<br>
<br>
2. According to Second Substitute Bill BL2018-1202 (as amended) (https://web.archive.org/web/20181019234657/https://www.nashville.gov/Metro-Clerk/Legislative/Ordinances/Details/7d2cf076-b12c-4645-a118-b530577c5ee8/2015-2019/BL2018-1202.aspx), all permitted operators will first clean data before providing or reporting data to Metro. Data processing and cleaning shall include:<br>
-- Removal of staff servicing and test trips<br>
-- Removal of trips below one minute<br>
-- Trip lengths are capped at 24 hours<br>
Are the scooter companies in compliance with the second and third part of this rule?
<br>
<br>
3. The goal of Metro Nashville is to have each scooter used a minimum of 3 times per day. Based on the data, what is the average number of trips per scooter per day? Make sure to consider the days that a scooter was available. How does this vary by company?
<br>
<br>
4. Metro would like to know how many scooters are needed, and something that could help with this is knowing peak demand. Estimate the highest count of scooters being used at the same time. When were the highest volume times? Does this vary by zip code or other geographic region?
<br>
<br>
5. Stretch Goal: SUMDs can provide alternative transportation and provide "last mile" access to public transit. How often are trips starting near public transit hubs? You can download a dataset of bus stop locations from https://data.nashville.gov/Transportation/WeGo-Transit-Bus-Stops/vfe9-k7vc/about_data.

#### Q1:

In [ ]:
#How many scooters did each company have
query = '''SELECT companyname, count(DISTINCT(sumdid))
    FROM scooters
    GROUP BY companyname'''

with engine.connect() as connection:
    company_count = pd.read_sql(text(query), con = connection)
    
company_count.sort_values('count', ascending=False)

In [28]:
#Did the number for each company change over time?
query = '''SELECT DISTINCT
                sumdid,
                companyname,
                EXTRACT(MONTH from pubdatetime) AS month,
                EXTRACT(DAY from pubdatetime) AS day
            FROM scooters;'''
with engine.connect() as connection:
    scooters_over_time = pd.read_sql(text(query), con = connection)
    
scooters_over_time['companyname'].value_counts()

companyname
Bird      134639
Jump       77724
Lime       77265
Lyft       56230
Spin       34469
Gotcha     16762
Bolt       16474
Name: count, dtype: int64

In [ ]:
#Did scooter usage vary by company?

#### Q2:

In [36]:
#import relevant trip duration data
tripdur = pd.read_csv('../data/trips.csv')

#drop irrelevant columns
tripdur = tripdur.drop(columns=['sumdid','tripdistance','startlatitude','startlongitude','endlatitude','endlongitude','triproute','create_dt'])
#print(tripdur.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 565522 entries, 0 to 565521
Data columns (total 16 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   pubtimestamp    565522 non-null  object 
 1   companyname     565522 non-null  object 
 2   triprecordnum   565522 non-null  object 
 3   sumdid          565522 non-null  object 
 4   tripduration    565522 non-null  float64
 5   tripdistance    565522 non-null  float64
 6   startdate       565522 non-null  object 
 7   starttime       565522 non-null  object 
 8   enddate         565522 non-null  object 
 9   endtime         565522 non-null  object 
 10  startlatitude   565522 non-null  float64
 11  startlongitude  565522 non-null  float64
 12  endlatitude     565522 non-null  float64
 13  endlongitude    565522 non-null  float64
 14  triproute       565522 non-null  object 
 15  create_dt       565522 non-null  object 
dtypes: float64(6), object(10)
memory usage: 69.0+ MB
None
<c

In [55]:
#transform data

tripdur.pubtimestamp = pd.to_datetime(tripdur.pubtimestamp)
tripdur.startdate = pd.to_datetime(tripdur.startdate)
tripdur.starttime = pd.to_datetime(tripdur.starttime)#, format = '%I:%M:%S:%f')
tripdur.enddate = pd.to_datetime(tripdur.enddate)
tripdur.endtime = pd.to_datetime(tripdur.endtime)#, format = '%I:%M:%S:%f')

#combine datetimes and convert to datetime objects

tripdur['tripstart'] = tripdur['startdate'].dt.strftime('%Y-%m-%d') + ' ' + tripdur['starttime'].dt.strftime('%H:%M:%S')
tripdur['tripend'] = tripdur['enddate'].dt.strftime('%Y-%m-%d') + ' ' + tripdur['endtime'].dt.strftime('%H:%M:%S')

tripdur.tripstart = pd.to_datetime(tripdur.tripstart)
tripdur.tripend = pd.to_datetime(tripdur.tripend)

print(tripdur.dtypes)

pubtimestamp     datetime64[ns]
companyname              object
triprecordnum            object
tripduration            float64
startdate        datetime64[ns]
starttime        datetime64[ns]
enddate          datetime64[ns]
endtime          datetime64[ns]
tripstart                object
tripend                  object
dtype: object
pubtimestamp     datetime64[ns]
companyname              object
triprecordnum            object
tripduration            float64
startdate        datetime64[ns]
starttime        datetime64[ns]
enddate          datetime64[ns]
endtime          datetime64[ns]
tripstart        datetime64[ns]
tripend          datetime64[ns]
dtype: object


In [70]:
#calculate trip durations
tripdur['duration_full'] = tripdur['tripend'] - tripdur['tripstart']
tripdur.dtypes
tripdur['trip_min'] = tripdur.duration_full / dt.timedelta(minutes=1)

In [134]:
#define compliance variables

#Trips Less Than One Minute
tripdur_sub1min = tripdur[tripdur['trip_min'] < 1]

#Trips Greater than 24 Hours
tripdur_over24h = tripdur[tripdur['trip_min'] > 1440]

#Noncompliant Trips
tripdur_noncompliant = tripdur_sub1min._append(tripdur_over24h, ignore_index=True)

#noncompliant companies
noncompliant_companies = tripdur_noncompliant['companyname'].unique()



array(['Bird', 'Lyft', 'Lime', 'Bolt Mobility', 'SPIN', 'JUMP'],
      dtype=object)

In [ ]:
# DURATION ANALYSIS BY COMPNAY SUB MIN
(tripdur_sub1min
 .groupby(['companyname'])
 ['triprecordnum']
 .count()
 .reset_index()
 )

In [80]:
# DURATION ANALYSIS OVER 24 H
(tripdur_over24h
 .groupby(['companyname'])
 ['triprecordnum']
 .count()
 .reset_index()
 )

companyname  triprecordnum
0  Bolt Mobility              4
1           JUMP             27
2           Lyft              2
3           SPIN             28

In [123]:
# DURATION ANALYSIS ALL NONCOMPLIANT
(tripdur_noncompliant
 .groupby(['companyname'])
 ['triprecordnum']
 .count()
  .reset_index()
 .sort_values('triprecordnum', ascending=False)
 )

companyname  triprecordnum
0           Bird           6377
4           Lyft           4489
3           Lime           1250
1  Bolt Mobility            274
5           SPIN             28
2           JUMP             27

In [135]:
#totals

totaltrips = len(tripdur.index)
totalsub1min = len(tripdur_sub1min.index)
totalover24h = len(tripdur_over24h.index)
totalnoncompliant = totalsub1min + totalover24h
percentagenoncompliant = totalnoncompliant/totaltrips * 100

print('Total number of scooter trips reported: ' + str(totaltrips))
print('Total number of scooter trips with a duration less than one minute: ' + str(totalsub1min))
print('Total number of scooter trips with a duration over 24 hours: ' + str(totalover24h))

print('Total number of noncompliant scooter trips: ' + str(totalnoncompliant))
print('Percentage of noncompliant scooter trips: ' + str(percentagenoncompliant))

print('Companies with noncompliant scooter trip data: ' + str(noncompliant_companies))

Total number of scooter trips reported: 565522
Total number of scooter trips with a duration less than one minute: 12384
Total number of scooter trips with a duration over 24 hours: 61
Total number of noncompliant scooter trips: 12445
Percentage of noncompliant scooter trips: 2.2006217264757164
Companies with noncompliant scooter trip data: ['Bird' 'Lyft' 'Lime' 'Bolt Mobility' 'SPIN' 'JUMP']


#### A2: 
Total number of scooter trips reported: 565522 <br>
Total number of scooter trips with a duration less than one minute: 12384 <br>
Total number of scooter trips with a duration over 24 hours: 61 <br>
Total number of noncompliant scooter trips: 12445 <br>
Percentage of noncompliant scooter trips: 2.2006217264757164 <br>
Companies with noncompliant scooter trip data: ['Bird' 'Lyft' 'Lime' 'Bolt Mobility' 'SPIN' 'JUMP'] <br>

#### Q3: 

In [137]:
# The goal of Metro Nashville is to have each scooter used a minimum of 3 times per day. 
# Based on the data, what is the average number of trips per scooter per day? 

query = '''
SELECT DISTICNT
    t.sumdid as unit, 
    t.companyname
    t.COUNT(triprecordnum) as tripcount,
    s.COUNT(emilysmonthdaycolumn) as daysactive,
    t.COUNT(triprecordnum)/s.COUNT(emilysmonthdaycolumn) as avgtripsperday
  
FROM trips as t

LEFT JOIN scooters s
on t.sumdid = s.sumdid;'''

# Make sure to consider the days that a scooter was available. How does this vary by company?
groupby statement in python

SyntaxError: invalid syntax (1741639425.py, line 3)

#### Q4:

In [ ]:
# Metro would like to know how many scooters are needed, and something that could help with this is knowing peak demand. 
# Estimate the highest count of scooters being used at the same time. 
# When were the highest volume times? Does this vary by zip code or other geographic region?

